paperqa/llms.py

```diff
@@ -0,0 +1,169 @@
+import re
+from typing import Any, Callable, Coroutine, get_args, get_type_hints
+
+from openai import AsyncOpenAI
+
+from .prompts import default_system_prompt
+
+
+def guess_model_type(model_name: str) -> str:
+    import openai
+
+    model_type = get_type_hints(
+        openai.types.chat.completion_create_params.CompletionCreateParamsBase
+    )["model"]
+    model_union = get_args(get_args(model_type)[1])
+    model_arr = list(model_union)
+    if model_name in model_arr:
+        return "chat"
+    return "completion"
+
+
+def process_llm_config(llm_config: dict) -> dict:
+    """Remove model_type and try to set max_tokens"""
+    result = {k: v for k, v in llm_config.items() if k != "model_type"}
+    if "max_tokens" not in result or result["max_tokens"] == -1:
+        model = llm_config["model"]
+        # now we guess!
+        if model.startswith("gpt-4") or (
+            model.startswith("gpt-3.5") and "1106" in model
+        ):
+            result["max_tokens"] = 4096
+        else:
+            result["max_tokens"] = 2048  # ?
+    return result
+
+
+async def embed_documents(
+    client: AsyncOpenAI, texts: list[str], embedding_model: str
+) -> list[list[float]]:
+    """Embed a list of documents with batching"""
+    if client is None:
+        raise ValueError(
+            "Your client is None - did you forget to set it after pickling?"
+        )
+    response = await client.embeddings.create(
+        model=embedding_model, input=texts, encoding_format="float"
+    )
+    return [e.embedding for e in response.data]
+
+
+def make_chain(
+    client: AsyncOpenAI,
+    prompt: str,
+    llm_config: dict,
+    skip_system: bool = False,
+    system_prompt: str = default_system_prompt,
+) -> Callable[[dict, list[Callable[[str], None]] | None], Coroutine[Any, Any, str]]:
+    """Create a function to execute a batch of prompts
```

Comment:
Could you declare in the docstring that this make_chain function is a replacement for langchain in paperqa <=v41

I think this will help old users track this major change

Reply:
done